src/main/webapp/authorization.js

```diff
@@ -1,56 +1,124 @@
+  constructor() { // eslint-disable-line
+    GitHubAuthorizer.initializeFirebase();
+
+    /** @type {string} gitHub API token */
+    this.token = null;
+
+    /**
+     * the firbase instance associated with a given 
+     * GitHubAuthorizer
+     * @type {Firebase} 
+     */
+    this.firebase = firebase; // eslint-disable-line
   }
-}
 
-/**
- * Get firebase
- * @return {Firebase}
- */
-function getFirebase() {
-  return firebase; //eslint-disable-line
-}
+  /**
+   * The most recent access token for the GitHub API. To refresh this, the user
+   * must be signed out and signed in again. It is important to keep in mind
+   * that this token expires and must be used swiftly after calling
+   * toggleSignIn(). If authorization is yet to be toggled, this will return
+   * null. If an error occurs during authorization, it will again be set to
+   * null. This property will be set to null on sign out as well.
+   *
+   * NOTE: In the event that an access token times out, the token will still be
+   * available here and so, a non-null token is not a good sign of a valid
+   * token.
+   *
+   * @return {string} most recent token or null
+   */
+  getToken() {
+    return this.token;
+  }
 
+  /**
+   * The current user if authorized and null if not. For all intents and
+   * purposes, this function will serve as a true/false value in an if
+   * statement.
+   *
+   * @return {Firebase.User} current user or null
+   */
+  getUser() {
+    return this.firebase.auth().currentUser;
+  }
 
-/**
- * Function called when clicking the Login/Logout button.
- */
-function toggleSignIn() {
-  const firebase = getFirebase();
-  if (!firebase.auth().currentUser) {
-    let provider = new firebase.auth.GithubAuthProvider();
-    firebase.auth().signInWithPopup(provider).then(function(result) {
-      // This gives you a GitHub Access Token.
-      // You can use it to access the GitHub API.
-      const token = result.credential.accessToken; //eslint-disable-line
-      // The signed-in user info.
-      const user = result.user; //eslint-disable-line
-    }).catch(function(error) {
-      // Handle Errors here.
-      console.error(error);
-      // Change this in future iterations
-      alert(error.message);
-    });
-  } else {
-    // if logged in, sign out
-    firebase.auth().signOut();
+  /**
+   * Get firebase instance used with this object
+   * @return {Firebase} firebase instance
+   */
+  getFirebase() {
+    return this.firebase;
+  }
+
+  /**
+   * Function called when authorizing user to verify them as a gitHub user.
+   * This function doubles as an opt out when the current user wants to leave
+   * our site for good.
+   *
+   * Side Effects: when authorizing a user, it sets the current objects token
+   * to the returned GitHub API accessToken if a token is returned.
+   *
+   * NOTE: this function is asynchronous and should be used with an await
+   *
+   * @return {boolean} true if no errors occured during authorization, false
```

Comment:
I think it's slightly better practice to avoid using bools to signify success/error. True/false doesn't really capture all of the possibilities here, so it's better not to try to have it do that.

Maybe we can return the token from this function? It'll be valid if it succeeds, right, and null if not.

Reply:
This is my bad, I initially suggested this change to booleans. The one nuance with returning tokens is that there are no tokens to return if the users is being signed out, which means that errors and signing out become ambiguous in the return, hence why I suggested we return a boolean instead. Thoughts on this?

Reply:
Ah sorry I think I missed that, not your problem.

2 thoughts on this: 1, it might make more sense to just have two functions, one for logging in and one for logging out, and call them separately when we intend to. 2, you're right, the point is there are more than 2 outcomes of this function, either SIGN_IN_SUCCESS, SIGN_IN_FAIL, or SIGNOUT. I'd support having just a `var ATHENTICATION_STATUS = {}` with those 3 fields (basically an "enum" in Javascript https://stackoverflow.com/questions/287903/what-is-the-preferred-syntax-for-defining-enums-in-javascript). 

The enum would be nice to see in this PR if we decide to go down that route. A TODO to split this function into two is probably fine otherwise.

Reply:
Good point on the enum or splitting it up. Happy to see that enums are making an appearance. I used to use them a lot when I was making game engines in high school haha.

Reply:
done!